src/lib.rs

```diff
@@ -140,9 +141,104 @@ fn spawn_in_pool(jobs: Arc<Mutex<Receiver<Thunk>>>) {
     });
 }
 
+/// A scoped thread pool used to execute functions in parallel.
+///
+/// `ScopedThreadPool` is different from `TaskPool` in that:
+/// * It propagates panics that occur in the worker threads.
+/// * It doesn't require the `'static` bound on the functions that are executed.
+/// * Worker threads are joined when the `ScopedThreadPool` is dropped.
+///
+/// # Example
+///
+/// ```rust
+/// use threadpool::ScopedThreadPool;
+///
+/// let mut numbers: &mut [u32] = &mut [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
+///
+/// // We need an extra scope to shorten the lifetime of the pool
+/// {
+///     let pool = ScopedThreadPool::new(4);
+///     for x in &mut numbers[..] {
+///         pool.execute(move|| {
+///             *x += 1;
+///         });
+///     }
+/// }
+///
+/// assert_eq!(numbers, [2, 3, 4, 5, 6, 7, 8, 9, 10, 11]);
+/// ```
+pub struct ScopedThreadPool<'pool> {
+    sender: Sender<Thunk<'pool>>,
+    #[allow(dead_code)] // This field is only used to extend the lifetime of the guards
+    guards: Vec<JoinGuard<'pool, ()>>
+}
+
+impl<'pool> ScopedThreadPool<'pool> {
+    /// Spawns a new thread pool with `threads` threads.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if `threads` is 0.
+    pub fn new(threads: u32) -> ScopedThreadPool<'pool> {
+        assert!(threads >= 1);
+
+        let (sender, receiver) = channel();
+        let receiver = Arc::new(Mutex::new(receiver));
+
+        let mut guards = Vec::with_capacity(threads as usize);
+        for _ in 0..threads {
+            guards.push(spawn_scoped_in_pool(receiver.clone()));
+        }
+
+        ScopedThreadPool { sender: sender, guards: guards }
+    }
+
+    /// Executes the function `job` on a thread in the pool.
+    pub fn execute<F>(&self, job: F)
+        where F: FnOnce() + Send + 'pool
+    {
+        // Sender can only send objects with the `'static` lifetime bound, so we
+        // need to transmute the closure.
+        // This is safe because all threads are joined when the pool is dropped
+        // and a job will never outlive the 'pool lifetime.
+        self.sender.send(Box::new(job)).unwrap();
+    }
+}
+
+#[unsafe_destructor]
+impl<'a> Drop for ScopedThreadPool<'a> {
+    fn drop(&mut self) {
+        // We need to ensure that the sender is dropped before the JoinGuards
+        // Otherwise the threads will be joined and wait forever in the loop
+        let sender = mem::replace(&mut self.sender, channel().0);
```

Comment:
Yes. The performance hit of checking for `Some` pales in comparison when sending a message to another thread.